scene/3d/physics_body_3d.cpp

```diff
@@ -1305,22 +1319,28 @@ void CharacterBody3D::_move_and_slide_grounded(double p_delta, bool p_was_on_flo
 			gt.origin = gt.origin - result.travel;
 			set_global_transform(gt);
 
-			Vector3 motion_slide_norm = motion.slide(prev_floor_normal).normalized();
+			// Slide using the intersection between the motion plane and the floor plane,
+			// in order to keep the direction intact.
+			Vector3 motion_slide_norm = up_direction.cross(motion).cross(prev_floor_normal);
+			motion_slide_norm.normalize();
+
 			motion = motion_slide_norm * (motion_slide_up.length());
 			collided = true;
+		} else {
+			// No collision.
```

Comment:
If there is no collision, `last_motion` should be equals to `motion`. Seems to be empty here, you can remove this `else` condition to let the line `1343` do the job.